activerecord/lib/active_record/connection_adapters/schema_cache.rb

```diff
@@ -204,7 +204,16 @@ def deep_deduplicate(value)
             value.transform_keys { |k| deep_deduplicate(k) }.transform_values { |v| deep_deduplicate(v) }
           when Array
             value.map { |i| deep_deduplicate(i) }
-          when String, Deduplicable
+          when String
+            if RUBY_VERSION.start_with?('2.6') && (value.tainted? || !value.frozen?)
```

Comment:
Generally you don't want this type of version check on a hotpath. People on other Ruby version should pay the cost for this.

You can have something like:

```ruby
if RUBY_VERSION.start_with?('2.6')
  def deep_deduplicate(value)
    # version handling the bug
  end
else
  def deep_deduplicate(value)
    # normal version
  end
end

Reply:
Also I'm not sure about the `(value.tainted? || !value.frozen?)`. Based on the test your wrote, the bug happens when `value.tained? && !value.frozen?`. 

Tainted strings are rare enough that just checking `tainted?` should be enough I think.